src/Toml/Parser/TOML.hs

```diff
@@ -68,69 +68,58 @@ inlineTableP = between
     (text "{") (text "}")
     (tomlFromInline <$> hasKeyP `sepEndBy` text ",")
 
--- | Parser for a table.
-tableP :: Parser (Key, TOML)
-tableP = do
-    key  <- tableNameP
-    toml <- subTableContent key
-    pure (key, toml)
-
--- | Parser for an array of tables.
-tableArrayP :: Parser (Key, NonEmpty TOML)
-tableArrayP = do
-    key       <- tableArrayNameP
-    localToml <- subTableContent key
-    more      <- optional $ sameKeyP key tableArrayP
-    case more of
-        Nothing         -> pure (key, localToml :| [])
-        Just (_, tomls) -> pure (key, localToml <| tomls)
+
+-- | Parser for an array of tables under a certain key.
+tableArrayP :: Key -> Parser (NonEmpty TOML)
+tableArrayP key =
+    tableP (Just key) `NC.sepBy1` sameKeyP key tableArrayNameP
 
 -- | Parser for a '.toml' file
 tomlP :: Parser TOML
-tomlP = do
-    sc
-    (val, inline)  <- distributeEithers <$> many hasKeyP
-    (table, array) <- fmap distributeEithers
-        $ many
-        $ eitherPairP (try tableP) tableArrayP
-
-    pure TOML
-        { tomlPairs       = HashMap.fromList val
-        , tomlTables      = fromList $ inline ++ table
-        , tomlTableArrays = HashMap.fromList array
-        }
+tomlP = sc *> tableP Nothing <* eof
 
--- | Parser for full 'TOML' under a certain key
-subTableContent :: Key -> Parser TOML
-subTableContent key = do
+-- | Parser for a table under a certain key
+tableP :: Maybe Key -> Parser TOML
+tableP key = do
     (val, inline)  <- distributeEithers <$> many hasKeyP
     (table, array) <- fmap distributeEithers
         $ many
-        $ childKeyP key
-        $ eitherPairP (try tableP) tableArrayP
+        $ eitherPairP
+            (try $ do
+              (kDiff, k) <- childKeyP key tableNameP
```

Comment:
It looks like this parser can hit another corner case because it has backtracking due to nested `try` combinators inside self-recursive calls. My understanding is that you should put `try` as deep as possible. I see that `childKeyP` already has `try`. Would the tests still pass if you just remove that `try` from the line above (88)?

Reply:
I'll check!